### ISSUE TITLE
rename `fillIfUnfulfilled` to `fill`

### ISSUE THREAD

**Issue by jeremy-w** [bignerdranch/Deferred#19]
This is a "gotcha" in the API for now. It's common to have multiple producers for a deferred value (say, both a network operation AND possibly the user smacking a cancel button); having to remember to do `fillIfUnfulfilled` or risk exploding is a wart from my point of view.

**Comment by zwaldowski** [comment]
Ping @randomstep so he sees this conversation, pursuant to [his comment](https://github.com/bignerdranch/Deferred/pull/16#issuecomment-131473569).

**Comment by zwaldowski** [comment]
I'm not going to put words in anyone's mouths, but @jgallagher and I are together in disagreement about this. Common-case usage of `Deferred` itself (that is, not considering `Task` or Future/Promise) cases should assert if filled because it models being filled once-and-only once. In this way (w.r.t. @randomstep's "but don't load the metaphorical crashing gun"), the precondition in `Deferred` is more like Core Data threading assertions: silently eating the error would be obscuring programmer error and burying actual potential problems. Abstractions that truly do require racing behavior should take care to deactivate the assertion if that is correct for their abstraction.

**Comment by randomstep** [comment]
So ideally Deferred would only have a single owner / filler (something like what Rust does with the borrow checker). At least, it would be ideal if the 'once and only once' usage of fill could be enforced, rather than just documented. I seem to remember hearing some discussion of this previously.

So, putting aside the eventual API for a moment, Deferred has 2 uses:
1 - only a single provider will fill the Deferred, once and only once. If anyone else tries that is a semantic error.
2 - first come, first ... filled. Multiple providers can hold a reference to Deferred. Whoever fills it first wins. Everyone must check before filling, since someone else could get there first. Any later fills are just discarded.

`fill` is the current API for 1. `fillIfUnfilled` is the current API for 2. I think API 2 is clear. But these are really 2 completely different usages of Deferred. Would be great if the type system (or at least API) made clear that case 1 was different. I'm guessing that is more easily said than done though ... at a minimum I could imagine `fillOnlyOnce` being the natural corollary rather than `fill`.

**Comment by randomstep** [comment]
What if there was an `OwnedDeferred` that could only be owned by one instance at a time? I guess the OwnedDeferred (or a manager of said instance) would have to have a reference to the owner/filler, which could be updated when the owner wanted to pass ownership.

**Comment by zwaldowski** [comment]
Use case #2 is close to the kind of thing we want out of Task/Future/Promise. `OwnedDeferred` is not unreasonable, but does edge close enough to Task/Future/Promise. But the whole idea here is that `Deferred` is the pretty low-level building block for all this type of functionality. [Let Deferred be Deferred.](http://ak-hdl.buzzfed.com/static/enhanced/webdr01/2013/2/18/19/enhanced-buzz-9728-1361234344-0.jpg) When you need to use _just_ `Deferred`, you want it as it is — the [new API](https://github.com/bignerdranch/Deferred/pull/13/files#diff-84698bee4c3154d49155aa8e9a06b3a2R80) perhaps embodies this better.

**Comment by jeremy-w** [comment]
Both fill and fillIfUnfulfilled respect the value semantics of resolving a
deferred value to a single concrete value. Slamming on fillIfUnfulfilled in
a loop would be a performance error but would not affect the semantics of
the deferred in the least. Unlike Core Data, the underlying abstraction is
threadsafe, so there is no risk of silent corruption.

El sábado, 15 de agosto de 2015, Zachary Waldowski <notifications@github.com>
escribió:

> I'm not going to put words in anyone's mouths, but @jgallagher
> <https://github.com/jgallagher> and I are together in disagreement about
> this. Common-case usage of Deferred itself (that is, not considering Task
> or Future/Promise) cases should assert if filled because it models being
> filled once-and-only once. In this way (w.r.t. @randomstep
> <https://github.com/randomstep>'s "but don't load the metaphorical
> crashing gun"), the precondition in Deferred is more like Core Data
> threading assertions: silently eating the error would be obscuring
> programmer error and burying actual potential problems. Abstractions that
> truly do require racing behavior should take care to deactivate the
> assertion if that is correct for their abstraction.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bignerdranch/Deferred/issues/19#issuecomment-131477846>
> .
>


-- 
--
Jeremy W. Sherman <jeremy@bignerdranch.com>

**Comment by randomstep** [comment]
I am still stuck on the `fill` semantics not being clear from the API. I appreciate wanting the default to be crash early on bad use, rather than silently ignore (and potentially drop on the floor) later responses. However the crashing seems to me like it could be prone to race conditions, depending on whether I've unwound some stack / managed to cancel before something else attempts to fill.

Hypothetically, if I am using Deferred and it crashes due to trying to fill, how likely am I to fix it by switching from usage of `fill` to `fillIfUnfilled`? It seems likely commonly that would be the semantics I actually meant given that I've allowed multiple objects to fill. I might sometimes instead remove one of the fillers, but it seems to me that would be the less likely case.

So I'm still thinking that the user needs to opt into case 1 (and case 2) explicitly. Future/Promise (TFP) might give us case 2(?), but I was proposing `OwnedDeferred` to give us a clearer API for case 1. Or - did you mean to say that TFP would better meet case 1?

**Comment by zwaldowski** [comment]
> Hypothetically, if I am using Deferred and it crashes due to trying to fill, how likely am I to fix it by switching from usage of `fill` to `fillIfUnfilled`? 

100%. If you actually have a race, all parties filling should use `fillIfUnfilled`. That doesn't mean the default should change, that means that an abstraction should use the correct version for its abstraction.

**Comment by jeremy-w** [comment]
@zwaldowski: I'm confused how the assertion is needed to preserve the abstraction. Both `fillAndMaybeExplode` and `fillIfUnfulfilled` guarantee the fill-once behavior; we (quite rightly!) don't have a `shutUpAndFillWithThisThingNow` API that breaks the model.

One thing that's not clear to me is that we're all talking about the same abstraction. The abstraction I have in mind is just a value that might be unknown but is nevertheless a single value. Once it becomes determined, it will not change. We don't provide a way to change what the value has resolved to after the fact, so whether an API call blows up if the value has already been resolved seems orthogonal to the abstraction. If `fillIfUnfulfilled` does not return anything now, we could shift it to return a `didFill: Bool` value, and any client desiring the exploding behavior could `assert(deferred.fill(value), "hey, we should have been the ones to fill \(deferred)! What the heck?")`.

This lets a client recover the exploding behavior while simplifying the API to simply `fill`.

**Comment by zwaldowski** [comment]
Resolved in #38 and #29.
